File: src/main/kotlin/dev/epicsquid/thatsrad/entity/custom/SweepyEntity.java
package dev.epicsquid.thatsrad.entity.custom;

import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.entity.monster.Monster;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import software.bernie.geckolib3.core.AnimationState;
import software.bernie.geckolib3.core.IAnimatable;
import software.bernie.geckolib3.core.PlayState;
import software.bernie.geckolib3.core.builder.AnimationBuilder;
import software.bernie.geckolib3.core.controller.AnimationController;
import software.bernie.geckolib3.core.event.predicate.AnimationEvent;
import software.bernie.geckolib3.core.manager.AnimationData;
import software.bernie.geckolib3.core.manager.AnimationFactory;
import software.bernie.geckolib3.util.GeckoLibUtil;

public class SweepyEntity extends Mob implements IAnimatable {
    AnimationFactory manager = GeckoLibUtil.createFactory(this);
    protected SweepyEntity(EntityType<? extends Mob> pEntityType, Level pLevel) {
        super(pEntityType, pLevel);
    }
    //attributes
    public static AttributeSupplier setAttributes() {
        return Monster.createMonsterAttributes()
                .add(Attributes.MAX_HEALTH, 10.0D)
                .add(Attributes.ARMOR_TOUGHNESS, 1.0D)
                .add(Attributes.ATTACK_DAMAGE, 0.0D)
                .add(Attributes.ATTACK_SPEED, 1.0F)
                .add(Attributes.MOVEMENT_SPEED, 0.1F)
                .build();
    }

    //can collect items
    @Override
    public boolean canPickUpLoot() {
        return true;
    }

    //can hold items
    @Override
    public boolean canHoldItem(ItemStack stack) {
        return true;
    }

    //can transfer items into a chest
    @Override
    public boolean canBeLeashed(Player player) {
        return true;
    }

    
    //animation
    private <E extends IAnimatable> PlayState predicate(AnimationEvent<E> event) {
        if (event.isMoving()) {
            event.getController().setAnimation(new AnimationBuilder().addAnimation("animation.sweepy.walk", true));
            return PlayState.CONTINUE;
        }
        event.getController().setAnimation(new AnimationBuilder().addAnimation("animation.sweepy.idle", true));
        return PlayState.CONTINUE;
    }

    private PlayState attackPredicate(AnimationEvent event) {
        if (this.swinging && event.getController().getAnimationState().equals(AnimationState.Stopped)) {
            event.getController().markNeedsReload();
            event.getController().setAnimation(new AnimationBuilder().addAnimation("animation.sweepy.attack", false));
            this.swinging = false;
        }
        return PlayState.CONTINUE;
    }


    @Override
    public void registerControllers(AnimationData data) {
        data.addAnimationController(new AnimationController(this, "controller",
                0, this::predicate));
        data.addAnimationController(new AnimationController(this, "attackController",
                0, this::attackPredicate));
    }

    @Override
    public AnimationFactory getFactory() {
        return manager;
    }

}

File: src/main/kotlin/dev/epicsquid/thatsrad/entity/custom/SweepyEntity.kt
package dev.epicsquid.thatsrad.entity.custom

import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.Mob
import net.minecraft.world.entity.ai.attributes.AttributeSupplier
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.entity.monster.Monster
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import software.bernie.geckolib3.core.IAnimatable
import software.bernie.geckolib3.core.PlayState
import software.bernie.geckolib3.core.builder.AnimationBuilder
import software.bernie.geckolib3.core.builder.ILoopType
import software.bernie.geckolib3.core.controller.AnimationController
import software.bernie.geckolib3.core.event.predicate.AnimationEvent
import software.bernie.geckolib3.core.manager.AnimationData
import software.bernie.geckolib3.core.manager.AnimationFactory
import software.bernie.geckolib3.util.GeckoLibUtil

class SweepyEntity protected constructor(pEntityType: EntityType<out Mob?>?, pLevel: Level?) : Mob(pEntityType, pLevel),
    IAnimatable {
   private var factory = GeckoLibUtil.createFactory(this)

    //can collect items
    override fun canPickUpLoot(): Boolean {
        return true
    }

    //can hold items
    override fun canHoldItem(stack: ItemStack): Boolean {
        return true
    }

    //can transfer items into a chest
    override fun canBeLeashed(player: Player): Boolean {
        return true
    }

    //animation
    private fun <E : IAnimatable?> predicate(event: AnimationEvent<E>): PlayState {
        if (event.isMoving) {
            event.controller.setAnimation(AnimationBuilder().addAnimation("animation.sweepy.walk", ILoopType.EDefaultLoopTypes.LOOP))
            return PlayState.CONTINUE
        }
        event.controller.setAnimation(AnimationBuilder().addAnimation("animation.sweepy.idle", ILoopType.EDefaultLoopTypes.LOOP))
        return PlayState.CONTINUE
    }
    override fun registerControllers(data: AnimationData) {
        data.addAnimationController(
            AnimationController(
                this,
                "controller",
                0f
            ) { event: AnimationEvent<SweepyEntity> -> predicate(event) })
    }

    override fun getFactory(): AnimationFactory {
        return factory
    }

    companion object {
        //attributes
        fun setAttributes(): AttributeSupplier {
            return Monster.createMonsterAttributes()
                .add(Attributes.MAX_HEALTH, 10.0)
                .add(Attributes.ARMOR_TOUGHNESS, 1.0)
                .add(Attributes.ATTACK_DAMAGE, 0.0)
                .add(Attributes.ATTACK_SPEED, 1.0)
                .add(Attributes.MOVEMENT_SPEED, 0.1)
                .build()
        }
    }
}
